ext/nnstreamer/tensor_filter/tensor_filter_caffe2.c

```diff
@@ -0,0 +1,192 @@
+/**
+ * GStreamer Tensor_Filter, caffe2 Module
+ * Copyright (C) 2019 Hyoung Joo Ahn <hello.ahn@samsung.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ */
+/**
+ * @file	tensor_filter_caffe2.c
+ * @date	27 May 2019
+ * @brief	Caffe2 for tensor_filter gstreamer plugin
+ * @see		http://github.com/nnsuite/nnstreamer
+ * @author	HyoungJoo Ahn <hello.ahn@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * This is the per-NN-framework plugin (caffe2) for tensor_filter.
+ * Fill in "GstTensorFilterFramework" for tensor_filter.h/c
+ *
+ */
+
+#include <glib.h>
+#include <string.h>
+
+#include "tensor_filter_caffe2_core.h"
+
+void init_filter_caffe2 (void) __attribute__ ((constructor));
+void fini_filter_caffe2 (void) __attribute__ ((destructor));
+
+/**
+ * @brief internal data of caffe2
+ */
+struct _Caffe2_data
+{
+  void *caffe2_private_data;
+};
+typedef struct _Caffe2_data caffe2_data;
+
+
+/**
+ * @brief Free privateData and move on.
+ */
+static void
+caffe2_close (const GstTensorFilterProperties * prop, void **private_data)
+{
+  caffe2_data *cf2;
+  cf2 = *private_data;
+  caffe2_core_delete (cf2->caffe2_private_data);
+  g_free (cf2);
+  *private_data = NULL;
+}
+
+/**
+ * @brief Load caffe2 modelfile
+ * @param prop property of tensor_filter instance
+ * @param private_data : caffe2 plugin's private data
+ * @return 0 if successfully loaded. 1 if skipped (already loaded).
+ *        -1 if the object construction is failed.
+ *        -2 if the object initialization if failed
+ */
+static int
+caffe2_loadModelFile (const GstTensorFilterProperties * prop,
+    void **private_data)
+{
+  caffe2_data *cf2;
+  if (*private_data != NULL) {
+    /** @todo : Check the integrity of filter->data and filter->model_file, nnfw */
+    cf2 = *private_data;
+    if (g_strcmp0 (prop->model_file,
+            caffe2_core_getModelPath (cf2->caffe2_private_data)) != 0 ||
+        g_strcmp0 (prop->model_file_sub,
+            caffe2_core_getInitModelPath (cf2->caffe2_private_data)) != 0) {
+      caffe2_close (prop, private_data);
+    } else {
+      return 1;
+    }
+  }
+  cf2 = g_new0 (caffe2_data, 1); /** initialize cf2 Fill Zero! */
```

Comment:
Is cf2 freed after returning error?

Reply:
`return 1` doesn't mean an error. It means maintaining current state rather than closing&opening new caffe2 with the new models.

Reply:
@helloahn Do you mean that cf2 would be freed after all in the cases of returning -2 and -1?

Reply:
@wooksong 
I did some tests for that case, and you're right. It's not freed if there is any problem. Specifically, `.close()` is not called.
Furthermore, our nnfws(like tensorflow, tf-lite, pytorch..) have the same problem. I'll fix it at once at another pr.

Reply:
@helloahn 
> Furthermore, our nnfws(like tensorflow, tf-lite, pytorch..) have the same problem.
Right. It is need to be fixed later.

IMHO, in this case, simply add g_free just before ```return -1``` and ```return -2``` would be fixed. Anyway, not a critical issue at this point.